Supports array

kbrock opened the pull request:
## Overview

Change implementation of `supports` to not define methods.

1. https://github.com/ManageIQ/manageiq/issues/21989
2. https://github.com/ManageIQ/manageiq/issues/21179

## Before

1. `supports :feature` and `supports_not :feature` defines `supports_{feature}?`.
2. if the feature is not supported, `supports_{feature}?` calls `add_unsupported_reason` to store the reason.
3. `supports?(feature)` calls `supports_{feature}?` under the covers.
4. `supports?(unknown_feature)` detects when `supports_{unknown_feature}?` is not defined and directly calls `add_unsupported_reason`.

## After

1. `supports :feature` and `supports_not :feature` stores the reasons in a class level hash.
2. if the feature is not supported, `supports_{feature}?` calls `add_unsupported_reason` to store the reason.
3. `supports?(feature)` uses class level hash to determine the reason. It leverages `add_unsupported_reason` the same as before.
4. `supports?(unknown_feature)` notices that the feature is not in the class level hash and directly calls `add_unsupported_reason`.

## Highlights

### Supports Attribute

We have a feature called `supports_attributes` columns that introduces a virtual attribute `supports_{feature}` (note: no trailing question mark). Since this is a boolean attribute, Rails defines this method with a question mark, which conflicts with the method defined in supports (before 1)

So when we try and determine if the feature is defined (before 4), it always detects true, and goes through rails attribute accessors. Rails 6.1 worked fine, but rails 7 changed the implementation and broke.

So this implementation needs to change

### dropping supports_feature? methods

The `supports_{feature}?` method is useful when using `detect` or `select` -- `detect(&:supports_feature?)`, so we used it a few times in the ui-classic and in core. We have pruned these out but unfortunately we do loose this shortcut. I think we must have introduced other methods that make this shortcut no longer needed because I could not find any `select.*supports\?` or `detect.*supports\?` methods.

### Determining if supports_feature? methods are used

Searching the code base for calls to `supports_feature?` is a little tricky since there are methods with the name `supports_feature?` defined, but they are not using the `supports :feature` mechanism. There is an initiative (linked at the top) to converge on the `supports :feature` mechanism, but that has been a background thread

### reducing the number of methods defined

I remember having a conversation with @chessbyte to cut down on the number of methods defined in classes, but I can't find an issue actually stating this goal. That was the main reason I had embarked on this journey some dozen PRs ago.

We have not completed to goal to converge all the supports type mechanisms.
But this PR completes the goal to reduce all the extra methods defined on the class.

We did reduce quite a few methods defined by #22578 - that no longer defines the methods for every base class. That PR definitely exposed the problem we are seeing now, but supports should not be going through the rails attributes code, so that is why I am suggesting this solution

kbrock commented:
ok, `supports_order?` is being used, though I think it was introduced in https://github.com/ManageIQ/manageiq/pull/19186 / https://github.com/ManageIQ/manageiq-api/pull/656

I'll see if that is used anywhere. Since these sometimes get exposed as columns, 

WIP: possibly drop supports_order / validate_supports
WIP: want to discuss before merging (and run cross tests

jrafanie commented:
FYI, I tried this PR with my rails 7 branch and things look good.

Fryguy commented:
The only thing I can think of that I didn't see covered in any of the comments here (unless I missed it), is that we have a handful of virtual columns representing "supports", which are meant to be accessed over the API.  As long as those still work, I'm good with this.

kbrock commented:
@Fryguy 

Thanks I changed Highlights / Supports Attribute to better highlight `attribute_supports`. I think this is what caused the issue in the first place.

kbrock commented:
update:

- dropped aliases to orderable? (PR in ui-classic to drop)
- dropped validate - not able to find this anywhere

update:

- added validate_order - looks like it may be called by `order_#{action}` from ui-classic

kbrock commented:
update:

- removed the deletion of ServiceTemplate#orderable? (OrchestrationTemplate implements this and this is no longer a quick change - we can do this work in a separate PR if we even want to do this)

Fryguy commented on app/models/mixins/supports_feature_mixin.rb:
```diff
@@ -51,18 +47,16 @@
 
-  def self.reason_or_default(reason)
-    reason.present? ? reason : _("Feature not available/supported")
+  def self.reason_or_default(reason = nil)
+    reason.presence || _("Feature not available/supported")
```
I'm not sure why this method should take a parameter at all - the caller can take care of it in all cases (and it only seems to be used here and in support_helper?

Thinking we should change this to more simply:
```suggestion
  def self.default_reason
    _("Feature not available/supported")
```

Then just change the callers from `SupportsFeatureMixin.reason_or_default(reason)` to `reason.presence || SupportsFeatureMixin.default_reason` or perhaps `reason ||= SupportsFeatureMixin.default_reason`

kbrock commented:
update:

- simplified default_supports_reason
- no longer setting a default reason keeping false value. that way it can be localized later
- added specs that chain supports (based upon the current code examples)

Fryguy commented on app/models/mixins/supports_feature_mixin.rb:
```diff
@@ -99,24 +87,46 @@ def unsupported
+      check_supports(feature.to_sym, :instance => self)
+    end
+
+    def check_supports(feature, instance:)
```
super-minor but is this intended to be a public or private interface?  If private, it should probably be marked as such.

kbrock replied:
it is private. thanks

kbrock replied:
Yes, in the next pr we drop this method.

Fryguy commented on app/models/mixins/supports_feature_mixin.rb:
```diff
@@ -99,24 +87,46 @@ def unsupported
     # This is the DSL used a class level to define what is supported
     def supports(feature, &block)
-      define_supports_feature_methods(feature, &block)
+      self.supports_features = supports_features.merge(feature.to_sym => block || true)
```
My only concern here is that I think this needs to be made thread-safe.  This is because we could lazy load classes at run-time in different threads (think 2 different requests coming that each need to lazy load things).

I know that lazy loading has bit us in the past with this - see the way the `unsupported` method is defined to use a Concurrent::Hash for this reason.

kbrock replied:
This is the Rails pattern for defining class level configuration that inherits from parent classes and then inherit columns from parents and allow overriding.

- define class variable https://github.com/rails/rails/blob/7-0-stable/activerecord/lib/active_record/attributes.rb#L11
- append to class variable https://github.com/rails/rails/blob/7-0-stable/activerecord/lib/active_record/attributes.rb#L231
- use variable to set methods https://github.com/rails/rails/blob/7-0-stable/activerecord/lib/active_record/attributes.rb#L265

We use this same pattern for virtual_attributes:

- https://github.com/kbrock/activerecord-virtual_attributes/blob/master/lib/active_record/virtual_attributes.rb#L41
- https://github.com/kbrock/activerecord-virtual_attributes/blob/master/lib/active_record/virtual_attributes.rb#L62


Since these values are set at class load time, and Rails has a class loader mutex to protect the other loaders, I am pretty sure we are in the clear.

kbrock replied:
`supported_features`, similar to rails attribute API, only modify the class at load time and should be protected by the rails class loader mutex.

`unsupported` is constantly modified at runtime. While I am not happy with this race condition, this PR focuses on removing the extra methods that were causing the infinite backtraces. In the future we can hopefully drop `unsupported` and use `supported_feature` instead.

kbrock replied:
I created #22898 as a place holder for the next phase

Fryguy replied:
This code does a get->modify->set, and another thread can get in the middle.  In the past, 2 threads could lazy-load 2 separate classes, particularly in the API.

> only modifies the class at load time and should be protected by the rails class loader mutex.

@jrafanie Can you comment here?  Not sure if this is now completely handled by zeitwerk or not, but this was a problem in the past and is why we had to wrap all of those class loading places in the special Rails handler, particularly in the API.

jrafanie replied:
hmm, there is [discussion of thread safety for code reloading](https://github.com/fxn/zeitwerk?tab=readme-ov-file#thread-safety).  I'm wondering if we need to implement something similar to [what we had previously](https://github.com/ManageIQ/manageiq/pull/22801) for dev mode reloading only.  I think autoloading in production mode is fine as long as it's after the application is initialized/rails loaded.

kbrock replied:
> This code does a get->modify->set, and another thread can get in the middle. In the past, 2 threads could lazy-load 2 separate classes, particularly in the API.

Yes, The 'get, modify, set' of `unsupported[]` is bad. This is the existing behavior. 

The first commit's goal is to just consolidate the existing logic.
Rails uses this patter when defining methods: `def field_name ; attr_reader(:field_name) ; end` instead of putting the actual gook into the defined method. (not always, but generally) 

```diff
  def supports(feature, &block)
    define_method("supports_#{feature}?") do
+     common_method(feature)
+   end
+ end
+
+   def common_method(feature) # sorry rubocop for indentation
      unsupported.delete(feature)
      if block
        instance_eval(&block) # gook (not for the class level)
      else
        unsupported.add(feature) # some kind of modify
      end
      !!unsupported[feature]
    end
- end
```

I can see if there is a way to make this fact more obvious.
But again, I wasn't as concerned about fixing the code knowing that #22898 will remove the `clear ; modify ; get` to a just plain `get`.


>> only modifies the class at load time
>> [...] this was a problem in the past and is why we had to wrap all of those class loading places in the special Rails handler

There have been issues when one class referenced another class at load time. `virtual_delegates` used to have this problem but we changed it to lazy reference the other class. `has_many` also has this problem, and why we use `class_name=String` rather than class references.

In our case, we are following the example from the `attributes()` api. It defines Symbols/Strings that are put into a `Hash` at load time and never changed after that.

Fryguy replied:
Ok discussed with @kbrock and I was mistakenly thinking that the class_attribute would be _shared_, so that child classes would be pushing values and could clobber each other.  However, what's actually happening is the parent class loads first, then each child creates a copy of the parent class' feature and then inserts their own into the copy.

So this is fine :+1:

kbrock commented:
update:
- rebase (no conflicts)
- merged check_supports into class level supports?

kbrock commented:
update:
- rebased (resolved conflict - supports?(:order) fix was fixed by #22902)

kbrock commented:
WIP: I found some `supports_feature?` methods in a few providers

kbrock commented:
searching for supports_:

docs to invalid methods:

- [ ] supports_cockpit (docs)
- [ ] supports_console (docs) -- need to expose?
- [ ] supports_vnc_console (docs)
- [ ] javascript grep decorators=vms.supports_*?


ignore:

- tools supports_nothing?
- inventory_refresh supports_building_inventory_collection?

currently defined: (and handled in part of https://github.com/ManageIQ/manageiq/issues/21989 )

- supports_sti?
- supports_pxe?
- supports_iso?
- supports_cloud_init?
- supports_sysprep?
- supports_customization_template?
- supports_systemd?
- supports_nohup_and_backgrounding
- supports_instance_store
- supports_ebs
- supports_security_groups ?
- supports_button_action
- supports_native_clone

columns:

- supports_32_bit?
- supports_64_bit?
- supports_hvm?
- supports_paravirtual?

methods with attributes

- supports_column?(v)
- supports_remote_data_timestamp?(v)
- supports_remote_data_version?(v)
- supports_resource_version?(v)
- supports_authentication(v)
- supports_command
- supports_migrate_for_all

kbrock commented:
un-WIP - there should be no `supports_feature?` methods defined

@miq-bot @miq-bot cross_repo_tests /all

kbrock commented:
@miq-bot cross_repo_tests #22883 /all

kbrock commented:
cross repo tests: automate and kubernetes failures are due to no merge yet

kbrock commented:
update:
- rebased (to pull in agrare's change around managers - to turn kubernetes green)

kbrock commented:
@Fryguy You have any concerns dropping the `supports_feature?` syntax from our code base?

Fryguy commented:
Nope I love it

kbrock commented:
update:
- rebase (fix merge conflicts)
